[doc] Update Rust for C devs with new alternatives and fixed links (#29675)
lowRISC/opentitan: doc/rust_for_c_devs.md

When pointers are dereferenced, they must be well-aligned and point to valid memory, like in C; failure to do so is UB.
Unlike in C, the address-of operator, `&x`, produces a reference[^33], rather than a pointer.
`&x as *const T` will create a pointer, instead.

> For creating pointers to values without creating an intermediate reference (which is safer, especially for unaligned or uninitialized memory), you should use the `addr_of!(x)` and `addr_of_mut!(x)` macros. See [https://doc.rust-lang.org/stable/core/ptr/macro.addr_of.html](https://doc.rust-lang.org/stable/core/ptr/macro.addr_of.html).

Reviewer: This has now become slightly out of date. `addr_of` and `addr_of_mut` are now superseeded by the `&raw` syntax.

```suggestion
Pointers are created with `&raw const x` (or `&raw mut x` for mutable pointers).
Older code may use the equivalent `addr_of` and `addr_of_mut` macros (which are part of `std::ptr` and are now deprecated) or `&x as *const T`. Using the latter is now discouraged as it creates a pointer via an intermediate reference which, in some cases, can lead to UB.
```

Rust macros are much more powerful than C macros, and operate on Rust syntax trees, rather than by string replacement.
Macro calls are differentiated from function calls with a `!` following the macro name.
For example, `file!()` expands to a string literal with the file name.

> To learn more about macros, see [https://veykril.github.io/tlborm/](https://veykril.github.io/tlborm/).

Reviewer: The link above now redirects to https://lukaswirth.dev/tlborm/
We should use the latter then.

It seems like a good idea to use this new version, which comes from an authoritative source (the rust-analyzer lead)

#### Inline Assembly

Rust supports inline assembly directly in the language.

> The syntax revolves around the stable macro `asm!()`, which integrates well with the language and is defined in the `core::arch` module.

Reviewer: Saying that it is stable and that it integrates well with the language is maybe superfluous.

```suggestion
Rust supports inline assembly via the macro `asm!()` defined in the `core::arch`module.
```

```
Unlike normal `let` statements, `if let` and `while let` expressions are meant to be used with refutable patterns.

> Rust also provides `let-else` statements, which allow you to use a refutable pattern with a normal `let` binding, provided you provide an `else` block that diverges (returns, panics, or breaks). This is extremely useful for early returns without adding indentation:

Reviewer: Nit: I would maybe simplify the phrasing a bit (your code snippet is quite useful and contain already quite a few of the details mentioned here). 

```suggestion
Rust also provides 'let-else' statements which are useful to implement early returns.
```
